crates/sui/src/client_commands.rs

```diff
@@ -3021,3 +3177,43 @@ async fn execute_dry_run(
         .map_err(|e| anyhow!("Dry run failed: {e}"))?;
     Ok(SuiClientCommandResult::DryRun(response))
 }
+
+/// Call a dry run with the transaction data to get the estimated gas budget.
+/// The final gas budget is the dry run gas budget + 10% margin.
+pub async fn estimate_gas_budget(
+    context: &mut WalletContext,
+    signer: SuiAddress,
+    kind: TransactionKind,
+    gas_price: u64,
+    gas_payment: Option<Vec<ObjectID>>,
+    sponsor: Option<SuiAddress>,
+) -> Result<u64, anyhow::Error> {
+    let client = context.get_client().await?;
+    let dry_run =
+        execute_dry_run(context, signer, kind, None, gas_price, gas_payment, sponsor).await;
+    if let Ok(SuiClientCommandResult::DryRun(dry_run)) = dry_run {
+        let buffer = GAS_SAFE_OVERHEAD * client.read_api().get_reference_gas_price().await?;
+        let gas_with_buffer = (dry_run.effects.gas_cost_summary().computation_cost + buffer) as i64;
+        let net_gas_usage = dry_run.effects.gas_cost_summary().net_gas_usage();
+        let gas_estimate = std::cmp::max(gas_with_buffer, net_gas_usage);
+        Ok(gas_estimate.try_into()?)
+    } else {
+        bail!("Dry run failed, could not automatically determine the gas budget. Please use the --gas-budget flag to provide a gas budget.")
+    }
+}
+
+/// Tries to get the maximum gas budget that is set in the protocol, and if not,
+/// it returns the max gas budget defined in a constant in this module {MAX_GAS_BUDGET}.
+pub async fn max_gas_budget(context: &mut WalletContext) -> Result<u64, anyhow::Error> {
+    let cfg = context
+        .get_client()
+        .await?
+        .read_api()
+        .get_protocol_config(None)
+        .await?;
+    Ok(match cfg.attributes.get("max_tx_gas") {
+        // TODO should we fail instead, if the protocol config does not have the max-budget, and ask the user to provide with a budget?
```

Comment:
I think it might actually be okay to fail in this case -- once a protocol config is set, it's not going to go away, so if we aren't getting a value here, that's the sign of some deeper issue, and the rest of the transaction is unlikely to succeed.